uninstall: remove symlinks created by flit install --symlink

sbidoul opened the pull request:
Fixes #6892 

This PR adds a failing test and a fix for uninstalling in a scenario that is very close to what `flit install --symlink` does.

The fix has two parts:
- avoid a redundant path normalization in `is_local` that was defeating careful *head* normalization in `req_uninstall.py`.
- avoid accidentally changing permission of the symlink target by considering symlinks as files instead of directories in `StashedUninstallPathSet`

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -420,17 +420,21 @@ def renames(old, new):
 
-def is_local(path):
-    # type: (str) -> bool
+def is_local(path, normalized=False):
```
Rather than complicating this function with a flag, and since I only count this function being called twice in the non-test code base, I would just document in the `:param:` that the path passed to the function must have had `normalize_path()` called on it, and then make sure that `normalize_path()` has been applied at each call site.

Also, I see that one of the two calls in the code base is directly passing the return value of `dist_location()`. Except for that one spot, it looks like every other call to `dist_location()` is calling `normalize_path()` on the return value of `dist_location()` anyways. So maybe either as part of this PR or as a separate PR, `dist_location()` can be changed to return a path that has had `normalize_path()` called on it. And that would make the code strictly simpler.

sbidoul commented:
@cjerdonek thanks for the review. Simplification done.

cjerdonek commented:
@sbidoul Thanks for the updates. Re: your comment [here](https://github.com/pypa/pip/issues/6892#issuecomment-522351100) on the original issue:

> But it reveals another issue: when uninstalling such symlinks, pip ends up modifying the permissions on the target directory to 0200...

Can you point out whether / how this PR addresses that? Is a code comment warranted if this was an issue that went unnoticed in the past?

sbidoul commented:
@cjerdonek I just extracted the part concerning stash/commit/rollback of symlinks in a separate commit and added extensive tests.

There is also an integration test for the mode change:
https://github.com/pypa/pip/blob/f8d58256b74ad1b8909a4efd86bdcc4449f54ef3/tests/functional/test_uninstall.py#L479

cjerdonek commented:
Sorry, I didn't mean to suggest this PR should (or needed) to address that comment. I was just asking because I wasn't sure. Does the second fix have to go with the first fix? In other words, would it cause problems if only the first part was done? If not, I think it would be better to do the second part as a separate issue and PR because it seems like a distinct issue and would make it easier to understand and review.

sbidoul commented:
I think both should be addressed together. Before this PR, uninstalling was merely leaving symlinks behind in the virtualenv. With the first part of this PR, the symlinks get removed but the stash mechanism corrupts the symlink targets. So I only merging the first part would make the situation worse and it's best to fix the stashing mechanism at the same time.

Is the separate commit sufficient to facilitate review or do you prefer a separate PR and merge them together?

sbidoul commented:
Also, the test for part 2 needs improvements in `tests/lib/path.py` that are in part one. So splitting the PR is somewhat more complicated.

cjerdonek commented:
> So I think only merging the first part would make the situation worse and it's best to fix the stashing mechanism at the same time.

Okay, this is good to know then. Before you had proposed only the first part without mentioning that more work needed to be done.

> Is the separate commit sufficient to facilitate review or do you prefer a separate PR and merge them together?

If both parts need to be committed together, then it should be one PR.

chrahunt commented on tests/lib/path.py:
```diff
@@ -203,9 +203,19 @@ def write_text(self, content):
         with open(self, "w") as fp:
             fp.write(content)
 
+    def append_text(self, content):
```
Let's put non-standard methods somewhere else.

chrahunt commented on src/pip/_internal/req/req_uninstall.py:
```diff
@@ -267,14 +267,18 @@ def _get_file_stash(self, path):
+        Handle symlinks as files to avoid modifying the symlink targets.
         """
-        if os.path.isdir(path):
+        if os.path.isdir(path) and not os.path.islink(path):
```
Can we assign this condition to something like `is_dir` and use it here and below?

chrahunt commented on src/pip/_internal/utils/misc.py:
```diff
@@ -427,10 +427,12 @@ def is_local(path):
 
     If we're not in a virtualenv, all paths are considered "local."
 
+    Caution: this function assumes the head of path has been normalized
```
Can we just normalize the head of path here and below?

sbidoul replied:
We could, but I think the callers need to do it too anyway, so for the use cases we have it would be redundant.

chrahunt replied:
Better redundant and give a consistent interface with no gotchas IMO, but I won't block this on this point.

chrahunt commented:
Thanks!
